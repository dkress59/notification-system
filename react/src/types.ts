import { DetailedHTMLProps, HTMLAttributes, PropsWithChildren } from 'react'

import { SpawnBannerProps, SpawnModalProps, SpawnToastProps } from ':core/types'

type CustomHTMLAttributes<
	T,
	K = void | Record<string, unknown>,
> = DetailedHTMLProps<HTMLAttributes<T> & K, T>

export type BannerAreaAttributes = CustomHTMLAttributes<HTMLBannerAreaElement>

export type NotificationAreaAttributes =
	CustomHTMLAttributes<HTMLNotificationAreaElement>

export type BannerNotificationAttributes =
	CustomHTMLAttributes<HTMLBannerNotificationElement>

export type ModalNotificationAttributes =
	CustomHTMLAttributes<HTMLModalNotificationElement>

export type ToastNotificationAttributes =
	CustomHTMLAttributes<HTMLToastNotificationElement>

declare global {
	// eslint-disable-next-line @typescript-eslint/no-namespace
	namespace JSX {
		interface IntrinsicElements {
			['banner-area']: BannerAreaAttributes
			['notification-area']: NotificationAreaAttributes
			['banner-notification']: BannerNotificationAttributes
			['modal-notification']: ModalNotificationAttributes
			['toast-notification']: ToastNotificationAttributes
		}
	}
}

export type BannerAreaProps = PropsWithChildren

export type NotificationAreaProps = PropsWithChildren

export interface BannerNotificationProps
	extends PropsWithChildren,
		Omit<SpawnBannerProps, 'content'> {
	onDismiss?: () => void
}

export interface ToastNotificationProps
	extends PropsWithChildren,
		Omit<SpawnToastProps, 'content'> {
	onDismiss?: () => void
}

export interface ModalNotificationProps
	extends PropsWithChildren,
		Omit<SpawnModalProps, 'content'> {
	onConfirm?: () => void
	onDecline?: () => void
	onDismiss?: () => void
}
