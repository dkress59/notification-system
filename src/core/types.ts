import {
	HTMLBannerAreaElement,
	HTMLBannerNotificationElement,
	HTMLModalNotificationElement,
	HTMLNotificationAreaElement,
	HTMLToastNotificationElement,
} from './components'

type SnakeToCamelCaseProp<S extends string> = S extends
	| `${infer T} ${infer U}`
	| `${infer T}-${infer U}`
	? `${Uncapitalize<T>}${Capitalize<SnakeToCamelCaseProp<U>>}`
	: Uncapitalize<S>
type SnakeToCamelCase<T> = {
	[K in keyof T as SnakeToCamelCaseProp<K & string>]: T[K]
}

export enum NotificationType {
	SUCCESS = 'success',
	INFO = 'info',
	ERROR = 'error',
	WARNING = 'warning',
}

export enum NotificationEvent {
	BANNER_DISMISSED = 'bannerDismissed',
	MODAL_CONFIRMED = 'modalConfirmed',
	MODAL_DECLINED = 'modalDeclined',
	MODAL_DISMISSED = 'modalDismissed',
	TOAST_DISMISSED = 'toastDismissed',
}

type SpawnArgs<T> = T & {
	content: string
}

export interface BannerAttributes {
	'auto-hide'?: boolean | string
	'auto-hide-after-ms'?: number | string
	'headline'?: string
	'type'?: NotificationType
}
export type SpawnBannerArgs = SpawnArgs<SnakeToCamelCase<BannerAttributes>>

export interface ModalAttributes {
	'condition'?: boolean
	'headline'?: string
	'show-confirm'?: boolean
	'show-decline'?: boolean
	'type'?: NotificationType
	'label-confirm'?: string
	'label-decline'?: string
}
export type SpawnModalArgs = SpawnArgs<SnakeToCamelCase<ModalAttributes>>

export interface ToastAttributes {
	'auto-hide'?: boolean | string
	'auto-hide-after-ms'?: number | string
	'headline'?: string
	'type'?: NotificationType
}
export type SpawnToastArgs = SpawnArgs<SnakeToCamelCase<ToastAttributes>>

declare global {
	interface HTMLElementTagNameMap {
		'banner-area': HTMLBannerAreaElement
		'banner-notification': HTMLBannerNotificationElement
		'modal-notification': HTMLModalNotificationElement
		'notification-area': HTMLNotificationAreaElement
		'toast-notification': HTMLToastNotificationElement
	}
}
