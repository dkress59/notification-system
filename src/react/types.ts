import { HTMLAttributes, MutableRefObject, PropsWithChildren } from 'react'

import {
	HTMLBannerAreaElement,
	HTMLBannerNotificationElement,
	HTMLModalNotificationElement,
	HTMLNotificationAreaElement,
	HTMLToastNotificationElement,
} from '../core/components'
import {
	BannerAttributes,
	ModalAttributes,
	SpawnBannerArgs,
	SpawnModalArgs,
	SpawnToastArgs,
	ToastAttributes,
} from '../core/types'

export type BannerAreaProps = PropsWithChildren & {
	forwardRef?: MutableRefObject<null | HTMLBannerAreaElement>
}

export type NotificationAreaProps = PropsWithChildren & {
	forwardRef?: MutableRefObject<null | HTMLNotificationAreaElement>
}

export interface BannerNotificationProps
	extends PropsWithChildren,
		Omit<SpawnBannerArgs, 'content'> {
	onDismiss?: () => void
	forwardRef?: MutableRefObject<null | HTMLBannerNotificationElement>
}

export type ToastNotificationProps = PropsWithChildren<
	Omit<SpawnToastArgs, 'content'> & {
		onDismiss?: () => void
		forwardRef?: MutableRefObject<null | HTMLToastNotificationElement>
	}
>

export interface ModalNotificationProps
	extends PropsWithChildren,
		Omit<SpawnModalArgs, 'content'> {
	onConfirm?: () => void
	onDecline?: () => void
	onDismiss?: () => void
	forwardRef?: MutableRefObject<null | HTMLModalNotificationElement>
}

type HtmlElementToReactNode<T = unknown> = HTMLAttributes<T> &
	PropsWithChildren<T> & {
		ref?: MutableRefObject<null | T>
	}

declare global {
	// eslint-disable-next-line @typescript-eslint/no-namespace
	namespace JSX {
		interface IntrinsicElements {
			['banner-area']: HtmlElementToReactNode
			['notification-area']: HtmlElementToReactNode
			['banner-notification']: HtmlElementToReactNode<BannerAttributes>
			['modal-notification']: HtmlElementToReactNode<ModalAttributes>
			['toast-notification']: HtmlElementToReactNode<ToastAttributes>
		}
	}
}
