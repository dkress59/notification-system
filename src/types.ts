import { PropsWithChildren } from 'react'

export enum ToastType {
	SUCCESS = 'success',
	INFO = 'info',
	ERROR = 'error',
	WARNING = 'warning',
}

export type FinalToastProps = PropsWithChildren<{
	autoHide?: boolean
	type?: ToastType
	title?: string | JSX.Element
	removeToastFromDom: () => void
}>
export type ToastProps = Omit<FinalToastProps, 'removeToastFromDom'>

export enum ToastAction {
	PUSH = 'push',
	POP = 'pop',
}

export interface ToastInCollection {
	id: string
	props: ToastProps
}
