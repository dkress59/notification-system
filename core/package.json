{
  "name": "core",
  "version": "0.0.1",
  "description": "Notification System Web Components",
  "main": "dist/index.cjs.js",
  "files": [
    "dist/bundle.js"
  ],
  "scripts": {
    "build": "webpack --mode production",
    "start": "webpack serve"
  },
  "devDependencies": {
    "autoprefixer": "10.4.7",
    "clean-webpack-plugin": "4.0.0",
    "css-loader": "6.7.1",
    "html-webpack-plugin": "5.5.0",
    "postcss": "8.4.14",
    "postcss-import": "14.1.0",
    "postcss-scss": "4.0.4",
    "sass": "1.52.1",
    "sass-loader": "13.0.0",
    "ts-jest": "28.0.2",
    "ts-loader": "9.3.0",
    "ts-node": "10.7.0",
    "webpack": "5.74.0",
    "webpack-cli": "4.10.0",
    "webpack-dev-server": "4.9.3"
  },
  "license": "MIT",
  "lint-staged": {
    "*.{scss, css}": [
      "yarn fix:style"
    ],
    "*.{js,jsx,ts,tsx}": [
      "yarn fix:script"
    ]
  }
}
