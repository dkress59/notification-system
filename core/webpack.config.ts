import 'webpack-dev-server'

import { CleanWebpackPlugin } from 'clean-webpack-plugin'
import HtmlWebpackPlugin from 'html-webpack-plugin'
import path from 'path'
import * as webpack from 'webpack'

interface WebpackArgv {
	env: {
		WEBPACK_BUNDLE: boolean
		WEBPACK_BUILD: boolean
		WEBPACK_SERVE: boolean
	}
	mode?: 'development' | 'production'
}

interface WebpackEnv {
	WEBPACK_BUNDLE: boolean
	WEBPACK_BUILD: boolean
	WEBPACK_SERVE: boolean
}

export default function webpackConfig(
	_env: WebpackEnv,
	argv: WebpackArgv,
): webpack.Configuration {
	const isProduction = argv.mode === 'production'
	const isBuild = !argv.env.WEBPACK_SERVE

	return {
		devServer: {
			compress: false,
			port: process.env.PORT ?? 3000,
		},

		devtool: isProduction ? false : 'inline-source-map',

		entry: './src/index.ts',

		mode: isProduction ? 'production' : 'development',

		module: {
			rules: [
				{
					test: /\.s?css?$/i,
					use: ['css-loader', 'sass-loader'],
					exclude: /node_modules/,
				},
				{
					test: /\.(t|j)sx?$/i,
					use: 'ts-loader',
					exclude: /node_modules/,
				},
			],
		},
		resolve: {
			extensions: ['.tsx', '.ts', '.js'],
		},

		output: isBuild
			? {
					path: path.resolve(__dirname, '..', 'dist'),
					publicPath: '',
					filename: 'bundle.js',
			  }
			: undefined,
		plugins: [
			new CleanWebpackPlugin(),
			isProduction
				? () => null
				: new HtmlWebpackPlugin({
						template: 'src/demo.html',
						filename: 'index.html',
				  }),
		],
	}
}
