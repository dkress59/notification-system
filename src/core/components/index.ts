export { HTMLBannerAreaElement } from './banner-area/banner-area'
export { HTMLBannerNotificationElement } from './banner-notification/banner-notification'
export { HTMLModalNotificationElement } from './modal-notification/modal-notification'
export { HTMLNotificationAreaElement } from './notification-area/notification-area'
export { HTMLToastNotificationElement } from './toast-notification/toast-notification'
