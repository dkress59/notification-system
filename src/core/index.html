<!DOCTYPE html>
<html dir="ltr" lang="en">

<head>
  <meta charset="utf-8" />
  <meta name="viewport" content="width=device-width, initial-scale=1.0, minimum-scale=1.0, maximum-scale=5.0" />
  <title>Notification Web Components</title>

  <script type="module" src="/bundle.js"></script>

  <style>
    * {
      box-sizing: border-box;
    }

    body {
      margin: 0;
      padding: 0;
      font-family: ui-sans-serif, system-ui, -apple-system, BlinkMacSystemFont, "Segoe UI", Roboto, "Helvetica Neue", Arial, "Noto Sans", sans-serif, "Apple Color Emoji", "Segoe UI Emoji", "Segoe UI Symbol", "Noto Color Emoji";
    }
  </style>
  <script>
    document.addEventListener('DOMContentLoaded', () => {
      document.querySelectorAll('toast-notification').forEach(toast => {
        toast.addEventListener('toastDismissed', (event) => console.debug(
          'toast dismissed.',
          { event }
        ))
      })
      document.querySelectorAll('modal-notification').forEach(modal => {
        modal.addEventListener('confirmTriggered', (event) => console.debug(
          'modal confirmed.',
          { event }
        ))
        modal.addEventListener('declineTriggered', (event) => console.debug(
          'modal declined.',
          { event }
        ))
      })

      const bannerArea = document.querySelector('banner-area')
      setTimeout(() => {
        bannerArea.spawnBanner({
          content: 'yoyoyo',
          autoHide: true,
        })
      }, 2000)
      setTimeout(() => {
        bannerArea.spawnBanner({
          content: 'yoyoyo',
          type: 'warning',
          headline: 'spawned',
        })
      }, 4000)

      const notificationArea = document.querySelector('notification-area')
      setTimeout(() => {
        notificationArea.spawnModal({
          content: 'yoyoyo',
          type: 'warning',
          headline: 'spawned',
        })
      }, 1500)
      setTimeout(() => {
        notificationArea.spawnModal({
          content: 'yoyoyo',
          headline: 'spawned new',
        })
      }, 4500)
    })
  </script>
</head>

<body>
  <banner-area>
    <banner-notification headline="some headline" type="error">demo content</toast-notification>
  </banner-area>
  <notification-area>
    <toast-notification class="test" headline="some headline" type="info">demo content</toast-notification>
  </notification-area>
</body>

</html>
