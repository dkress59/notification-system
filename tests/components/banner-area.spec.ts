import { screen } from '@testing-library/dom'

import { NotificationType } from ':core/types'

describe('<banner-area />', () => {
	beforeEach(() => {
		document.body.innerHTML = ''
	})
	it('matches snapshot', () => {
		const area = document.createElement('banner-area')
		area.setAttribute('data-testid', 'notification-area')
		area.innerHTML = '<p>slotted_paragraph</p>'
		document.body.appendChild(area)
		const element = screen.getByTestId('notification-area')
		expect(element.shadowRoot?.innerHTML).toMatchSnapshot()
	})
	it('can spawn a banner-notification (with defaults)', () => {
		const area = document.createElement('banner-area')
		area.setAttribute('data-testid', 'notification-area')
		document.body.appendChild(area)
		area.spawnBanner({ content: 'mock_content' })
		const element = screen.getByTestId('notification-area')
		expect(element.shadowRoot?.innerHTML).toMatchSnapshot()
	})
	it('spawned banner-notification receives props', () => {
		const mockProps = {
			content: 'mock_content',
			autoHide: true,
			autoHideAfterMs: 456,
			headline: 'mock_headline',
			type: NotificationType.INFO,
		}
		const area = document.createElement('banner-area')
		area.setAttribute('data-testid', 'notification-area')
		document.body.appendChild(area)
		area.spawnBanner(mockProps)
		const element = screen.getByTestId('notification-area')
		const notification = element.shadowRoot?.querySelector(
			'banner-notification',
		)
		expect(notification?.innerHTML).toEqual(mockProps.content)
		expect(notification).toHaveAttribute(
			'auto-hide',
			String(mockProps.autoHide),
		)
		expect(notification).toHaveAttribute(
			'auto-hide-after-ms',
			String(mockProps.autoHideAfterMs),
		)
		expect(notification).toHaveAttribute('headline', mockProps.headline)
		expect(notification).toHaveAttribute('type', mockProps.type)
	})
})
