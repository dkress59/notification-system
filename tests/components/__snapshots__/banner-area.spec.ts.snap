// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<banner-area /> can spawn a banner-notification (with defaults) 1`] = `
"<style>mock-style {
	mock-prop: value;
}</style><slot></slot><banner-notification aria-role="dialog" style="transition: none;" class="hidden">mock_content</banner-notification>"
`;

exports[`<banner-area /> matches snapshot 1`] = `
"<style>mock-style {
	mock-prop: value;
}</style><slot></slot>"
`;
